File: src/practice/party.c
// Author: Param Singh <paramsingh258@gmail.com>
// Party Schedule: http://www.spoj.com/problems/PARTY/
// Knapsack problem

#include <stdio.h>
#include <stdlib.h>

void knapsack(int* cost, int* fun, int w, int n) {
    int **dp   = (int **) calloc(n+1, sizeof(int*));
    int **keep = (int **) calloc(n+1, sizeof(int*));
    int i, j;
    for (i = 0; i <= n; i++) {
        dp[i] = (int *) calloc(w+1, sizeof(int));
        keep[i] = (int *) calloc(w+1, sizeof(int));
    }
    for (i = 1; i <= n; i++) {
        for (j = 0; j <= w; j++) {
            if ((cost[i-1] <= j) && (fun[i-1] + dp[i-1][j-cost[i-1]] > dp[i-1][j])) {
                keep[i][j] = 1;
                dp[i][j] = fun[i-1] + dp[i-1][j-cost[i-1]];
            }
            else {
                keep[i][j] = 0;
                dp[i][j] = dp[i-1][j];
            }
        }
    }

    int k = dp[n][w];
    int net;
    for (i = 0; i <= w; i++) {
        if (dp[n][i] == k) {
            net = i;
            break;
        }
    }

    printf("%d %d\n", net, dp[n][w]);
}

int main(void) {
    int w, n;
    while (scanf("%d%d", &w, &n) != EOF) {
        if (w == 0 && n == 0)
            break;
        int i;
        int *fun = (int *) calloc(n, sizeof(int));
        int *cost = (int *) calloc(n, sizeof(int));
        for (i = 0; i < n; i++)
            scanf("%d%d", cost+i, fun+i);
        knapsack(cost, fun, w, n);
    }
    return 0;
}

File: src/practice/party.py
def knapsack(cost, fun, w, n):
    dp = [[0 for _ in range(w+1)] for _ in range(n+1)]
    keep  = [[False for _ in range(w+1)] for _ in range(n+1)]
    for i in range(1, n+1):
        for j in range(w+1):
            if cost[i-1] <= j and (fun[i-1] + dp[i-1][j-cost[i-1]]) > dp[i-1][j]:
                keep[i][j] = True
                dp[i][j] = fun[i-1] + dp[i-1][j-cost[i-1]]
            else:
                keep[i][j] = False
                dp[i][j] = dp[i-1][j]

    net = 0
    k = w
    for i in range(n, 0, -1):
        if keep[i][k]:
            net += cost[i-1]
            k -= cost[i-1]

    print net, dp[n][w]

while True:
    raw_input()
    w, n = map(int, raw_input().split())
    if w == 0 and n == 0:
        break
    fun = [0 for _ in range(n)]
    cost = [0 for _ in range(n)]
    for i in range(n):
        cost[i], fun[i] = map(int, raw_input().split())
    knapsack(cost, fun, w, n)

